### DISCUSSION TITLE
How does one find which font kitty is using to render a particular character?

### DISCUSSION THREAD

**Discussion by lokesh-krishna** [kovidgoyal/kitty#3972] [Q&A] [answered]
I'm trying to symbol map a few Unicode characters to a font that has edited versions of it for use in a particular program (ncmpcpp) but while the edited characters show up in their edited form in other applications (GIMP, Charmap, etc), kitty still renders them as they were prior to editing and I'm not sure where it is getting them from. 

The characters in question are: U+2588, U+E0B6 and U+E0B4. I've edited them to be half of their height so that I can use them as progressbar characters in ncmpcpp. 

Here is how the characters (left-most and right-most) show up in Charmap:

![grim-2021-08-26-11-34-15](https://user-images.githubusercontent.com/20235646/130909581-61c482c2-467e-4801-8ad5-f3dbd9a9ce46.png)

Here is however how they are being rendered in kitty:

![grim-2021-08-26-11-35-40](https://user-images.githubusercontent.com/20235646/130909690-8e6f67a9-55f9-4fd7-8ab5-8e064060c941.png)

The fourth character is the cursor for reference that they are still full height even though Charmap shows them as being half the height. This is the case irrespective of whether these particular characters are symbol mapped to the font or the font is set as the default font family itself. 

How can I go about debugging it? 

Running `kitty --debug-font-fallback` returns the following:

```
[238 11:23:40.407369] Preloaded font faces:
[238 11:23:40.407407] normal face: /home/loki/.local/share/fonts/VictorMonoNerdFontComplete-Medium.otf:0
[238 11:23:40.407422] bold face: /home/loki/.local/share/fonts/VictorMonoNerdFontComplete-Medium.otf:0
[238 11:23:40.407439] italic face: /home/loki/.local/share/fonts/VictorMonoNerdFontComplete-Medium.otf:0
[238 11:23:40.407453] bi face: /home/loki/.local/share/fonts/VictorMonoNerdFontComplete-Medium.otf:0
```

VictorMonoNerdFontComplete is the edited font and if I'm reading that output right, it's the only font being used for any characters being rendered, correct? So I'm confused about where the full height characters for those particular characters are coming from.

**Comment by kovidgoyal** [comment, marked as answer]
kitty draws all box drawing characters programmatically, not from fonts, to guarantee they always align with cell edges. You cannot control how they are rendered.

**Reply by lokesh-krishna** [reply]
Thanks for the explanation! 
I moved my custom glyphs to some empty private use code points and they now render as expected. One doubt I have is why the alignment between two of the glyphs seems off and while I know this might entirely be on the glyphs themselves, the other two align perfectly so was wondering if you could point me in the right direction regarding why this might be happening. 

![grim-2021-08-26-13-16-03](https://user-images.githubusercontent.com/20235646/130922930-f9b3541b-d8e5-4e97-8f2e-bc7f234d2b39.png)

Referring to the alignment between the very first two glyphs of the progressbar here.

**Reply by kovidgoyal** [reply]
You cannot have font glyphs line up with 100% success, given the limitations of existing font tech. That is the whole reason kitty draws the box drawing characters itself in the first place. Some characters will work with some font sizes and some rendering engines some of the time. That is the best you can hope for.

**Reply by lokesh-krishna** [reply]
Thanks a lot for all your time :D

**Comment by lokesh-krishna** [comment]
Tried two different browsers but still not able to mark your reply as the answer. Can you please do it?

**Comment by kovidgoyal** [comment]
On Sun, Aug 29, 2021 at 04:29:31AM -0700, Lokesh Krishna wrote:
> Tried two different browsers but still not able to mark your reply as the answer. Can you please do it?

I believe github only lets one mark the top level reply as the answer.
